src/backend/src/user_profile.rs

```diff
@@ -0,0 +1,22 @@
+use crate::{Candid, StoredPrincipal, VMem};
+use ic_cdk::api::time;
+use ic_stable_structures::StableBTreeMap;
+use shared::types::user_profile::StoredUserProfile;
+
+pub fn get_or_create(
+    principal: StoredPrincipal,
+    user_profile_map: &mut StableBTreeMap<(u64, StoredPrincipal), Candid<StoredUserProfile>, VMem>,
+    user_profile_updated_map: &mut StableBTreeMap<StoredPrincipal, u64, VMem>,
+) -> Candid<StoredUserProfile> {
+    if let Some(updated) = user_profile_updated_map.get(&principal) {
+        user_profile_map
+            .get(&(updated, principal))
+            .expect("Failed to fetch user from user profile map but it's present in updated map")
+    } else {
+        let now = time();
+        let default_profile = StoredUserProfile::default(now);
```

Comment:
Not sure why we pass now as a parameter instead of scoping the creating of this timestamp?

Moreover, if we don't pass anything we can implement the default trait

```
impl Default for StoredUserProfile {
    fn default() -> Self {
```

Reply:
I wanted to keep the implementation of the class pure. `time` can only used within the canister, this way we could create one in a test for example.

What's the advantage of using the the default trait?

Reply:
Not a strong opinion but, I rarely create object by passing a timestamp as parameter.

i.e. I rarely did something like that

```
class Something {
   constructor(private now: number) {}
}

const tmp = new Something(Date.now());
```

I'm also a Rust noob but, have the feeling that generally `::default()` accept, is used with, no arguments.

For those reason, if you would not like to use the time in the implementation, maybe instead of using `default` we an implementation named `from` or `from_something` or `with_something`?

```
let default_profile = StoredUserProfile::from_timestamp(now);
```

That said, I can live with `default(now)` so not that a strong opinion neither. If you want to go with that for sure I'm fine with it.

Reply:
I agree with you on passing now. But I thought that in this case it would be beneficial because the "now" needs to be executed within a canister. Which means that if I put it in the class, then that class won't be usable in a test for example.